File: 1691-maximum-height-by-stacking-cuboids/1691-maximum-height-by-stacking-cuboids.py
class Solution(object):
    def maxHeight(self, cuboids):
        """
        :type cuboids: List[List[int]]
        :rtype: int
        """
        n = len(cuboids)
        for c in cuboids: c.sort()
        cuboids.sort()
        dp = [0] * n
        for i in range(n):
            dp[i] = cuboids[i][2]
            for j in range(0, i):
                if cuboids[j][0] <= cuboids[i][0] and cuboids[j][1] <= cuboids[i][1] and cuboids[j][2] <= cuboids[i][2]:
                    dp[i] = max(dp[i], dp[j] + cuboids[i][2])
        
        return max(dp)

File: 1691-maximum-height-by-stacking-cuboids/1691-maximum-height-by-stacking-cuboids.cpp
class Solution {
public:
    int maxHeight(vector<vector<int>>& cuboids) {
        int n = cuboids.size();
        for(auto& cub : cuboids) sort(cub.begin(), cub.end());
        sort(cuboids.begin(), cuboids.end());
        vector<int> dp(n);
        for(int i = 0; i < n; ++i) {
            dp[i] = cuboids[i][2];
            for(int j = 0; j < i; ++j) {
                if(cuboids[j][0] <= cuboids[i][0] && cuboids[j][1] <= cuboids[i][1] && cuboids[j][2] <= cuboids[i][2])
                    dp[i] = max(dp[i], dp[j] + cuboids[i][2]);
            }
        }
        
        int ret = 0;
        for(int d : dp) ret = max(ret, d);
        return ret;
    }
};